crates/ursa-consensus/src/abci_engine.rs

```diff
@@ -0,0 +1,189 @@
+use anyhow::{bail, Result};
+use std::net::SocketAddr;
+use tokio::sync::mpsc::Receiver;
+use tokio::sync::oneshot::Sender as OneShotSender;
+use tracing::warn;
+
+use narwhal_types::{Batch, Transaction};
+
+// Tendermint Types
+use crate::AbciQueryQuery;
+use tendermint_abci::{Client as AbciClient, ClientBuilder};
+use tendermint_proto::abci::{
+    RequestBeginBlock, RequestDeliverTx, RequestEndBlock, RequestInfo, RequestInitChain,
+    RequestQuery, ResponseQuery,
+};
+use tendermint_proto::types::Header;
+
+pub struct Engine {
+    /// The address of the ABCI app
+    pub app_address: SocketAddr,
+    /// Messages received from the ABCI Server to be forwarded to the engine.
+    pub rx_abci_queries: Receiver<(OneShotSender<ResponseQuery>, AbciQueryQuery)>,
+    /// The last block height, initialized to the application's latest block by default
+    pub last_block_height: i64,
+    pub client: AbciClient,
+    pub req_client: AbciClient,
```

Comment:
Whats the difference between these two?

Reply:
Queries will go through the req_client because they dont change state. Txns will go through client. The reason for having separate clients for this is because the client will maintain a lock while executing or committing transactions on the evm db. So having separate channels will not hold up queries when this is happening, and queries will return the last committed state, and not state in an inbetween state if its half way through executing.

Reply:
Having separate clients isn't necessary to separate reads from writes. Look at this client by aleo https://github.com/AleoHQ/aleo_lambda_blockchain/blob/main/src/client/tendermint.rs